keps/sig-scheduling/2249-pod-affinity-namespace-selector/README.md

```diff
@@ -0,0 +1,708 @@
+motivation section can optionally provide links to [experience reports] to
+demonstrate the interest in a KEP within the wider Kubernetes community.
+
+[experience reports]: https://github.com/golang/go/wiki/ExperienceReports
+-->
+
+### Goals
+- Allow users to dynamically select the set of namespaces considered when using pod
+  affinity/anti-affinity
+
+<!--
+List the specific goals of the KEP. What is it trying to achieve? How will we
+know that this has succeeded?
+-->
+
+<!--
+### Non-Goals
+
+What is out of scope for this KEP? Listing non-goals helps to focus discussion
+and make progress.
+-->
+
+## Proposal
+
+<!--
+This is where we get down to the specifics of what the proposal actually is.
+This should have enough detail that reviewers can understand exactly what
+you're proposing, but should not include things like API designs or
+implementation. What is the desired outcome and how do we measure success?.
+The "Design Details" section below is for the real
+nitty-gritty.
+-->
+
+### User Stories (Optional)
+
+<!--
+Detail the things that people will be able to do if this KEP is implemented.
+Include as much detail as possible so that people can understand the "how" of
+the system. The goal here is to make this feel real for users without getting
+bogged down.
+-->
+
+#### Story 1
+I am running a SaaS service, the workload for each customer is placed in a separate
+namespace. The workloads requires 1:1 pod to node placement. I want to use pod 
+anti-affinity across all customers namespaces to achieve that.
+
+<!--
+
+### Notes/Constraints/Caveats (Optional)
+
+What are the caveats to the proposal?
+What are some important details that didn't come across above?
+Go in to as much detail as necessary here.
+This might be a good place to talk about core concepts and how they relate.
+-->
+
+### Risks and Mitigations
+
+#### Performance
+Using namespace selector will make it easier for users to specify affinity 
+constraints across a large number of namespaces. The initial implementation 
+of pod affinity/anti-affinity suffered from performance challenges, however 
+over the releases 1.17 - 1.20 we significantly improved that. 
+
+We currently have integration test and clusterloader benchmarks that evaluate the extreme
+cases of all pods having affinity/anti-affinity constraints to/against each other.
+Those benchmarks show that the scheduler is able to achieve maximum throughput 
+(i.e., the api-server qps limit).
+
+#### Security/Abuse
+see previous discussion [here](https://github.com/kubernetes/kubernetes/issues/68827#issuecomment-468470598)
+NamespaceSelector will allow users to select all namespaces. This may cause a security 
+concern: a pod with anti-affinity constraint can block pods from all other namespaces 
+from getting scheduled in a failure domain.
+
+```
+<<[UNRESOLVED]>>
+We will partially address this concern by rejecting empty selectors and only
```

Comment:
lgtm. @liggitt ?

Reply:
I think this makes sense... perhaps a brief example of what could and could not be expressed with these restrictions would be helpful.

Reply:
sg, added.

Reply:
actual Equals and DoubleEquals don't make sense [1], the only one that does is "In", updated.

[1] https://github.com/kubernetes/kubernetes/blob/41a42ef6753a3fc2f90fa110c4a6dea0aa8304e9/staging/src/k8s.io/apimachinery/pkg/apis/meta/v1/types.go#L1114

Reply:
What about Exists?

Reply:
Problematic because namespaces will have common labels: https://github.com/kubernetes/enhancements/pull/2162, so users can easily target all namespaces as follows 

```
matchExpressions:
        - key: kubernetes.io/metadata.name
          operator: Exists
```